### PR TITLE
mgr/dashboard: Fix user/bucket count in rgw overview dashboard

### PR DESCRIPTION
**Pull request by aaSharma14**
Do not consider buckets/users count from daemons that have similar realm name

Fixes: https://tracker.ceph.com/issues/62964

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by cloudbehl — APPROVED** [review]
LGTM, Can you paste the recording on how it looks with the fix?

**Comment by aaSharma14** [comment]
jenkins test api

**Review by Pegonzal — COMMENTED** [review]
It would be nice to have a better description for the PR/commit, it is hard to follow what is changing here.

As far as I see, if a realm name is the same in another daemon, the `users_count` and the `buckets_count` will be set to the users/buckets of that daemon, but if it is not, it will be added up.

Without much context on the issue, it looks like I could have two realms and three daemon: like so: `realm1 and daemon1`, `realm1 and daemon2` and `realm2 and daemon3`, and I would end up with only the `users/buckets` count for the realm2, right?

**Comment by aaSharma14** [comment]
> It would be nice to have a better description for the PR/commit, it is hard to follow what is changing here.
> 
> As far as I see, if a realm name is the same in another daemon, the `users_count` and the `buckets_count` will be set to the users/buckets of that daemon, but if it is not, it will be added up.
> 
> Without much context on the issue, it looks like I could have two realms and three daemon: like so: `realm1 and daemon1`, `realm1 and daemon2` and `realm2 and daemon3`, and I would end up with only the `users/buckets` count for the realm2, right?

Hey @Pegonzal , The ask is to count the users/buckets per realm and not per daemon. As per your scenario -  I could have two realms and three daemon: like so: realm1 and daemon1, realm1 and daemon2 and `realm2 and daemon3... In this case it would end up with users/buckets count for realm1 only once(  i.e it will not count users/buckets for daemon1 and daemon2 twice) and will add up the user/bucket count for realm2. For more clarity its like - If I have daemon1 and daemon2 in realm1 and daemon3, daemon4 in realm2 , the result will be user/bucket count of daemon1 or daemon2 + user/bucket count of daemon3 or daemon4. The previous implementation added up daemon1+daemon2+daemon3+daemon4 which is misleading. Please let me know if it still confuses you, we can have a call.

**Comment by Pegonzal** [comment]
> Hey @Pegonzal , The ask is to count the users/buckets per realm and not per daemon. As per your scenario - I could have two realms and three daemon: like so: realm1 and daemon1, realm1 and daemon2 and `realm2 and daemon3... In this case it would end up with users/buckets count for realm1 only once( i.e it will not count users/buckets for daemon1 and daemon2 twice) and will add up the user/bucket count for realm2. For more clarity its like - If I have daemon1 and daemon2 in realm1 and daemon3, daemon4 in realm2 , the result will be user/bucket count of daemon1 or daemon2 + user/bucket count of daemon3 or daemon4. The previous implementation added up daemon1+daemon2+daemon3+daemon4 which is misleading. Please let me know if it still confuses you, we can have a call.

Thanks for clarifying a bit @aaSharma14, but I am still confused. 
Won't this code set the buckets and users to the last unique realm?
```
else:
                buckets = json.loads(RgwBucket.list(self, daemon_name=daemon['id']))
                users = json.loads(RgwUser.list(self, daemon_name=daemon['id']))
                users_count = len(users)
                buckets_count = len(buckets)
```

**Comment by aaSharma14** [comment]
> > Hey @Pegonzal , The ask is to count the users/buckets per realm and not per daemon. As per your scenario - I could have two realms and three daemon: like so: realm1 and daemon1, realm1 and daemon2 and `realm2 and daemon3... In this case it would end up with users/buckets count for realm1 only once( i.e it will not count users/buckets for daemon1 and daemon2 twice) and will add up the user/bucket count for realm2. For more clarity its like - If I have daemon1 and daemon2 in realm1 and daemon3, daemon4 in realm2 , the result will be user/bucket count of daemon1 or daemon2 + user/bucket count of daemon3 or daemon4. The previous implementation added up daemon1+daemon2+daemon3+daemon4 which is misleading. Please let me know if it still confuses you, we can have a call.
> 
> Thanks for clarifying a bit @aaSharma14, but I am still confused. Won't this code set the buckets and users to the last unique realm?
> 
> ```
> else:
>                 buckets = json.loads(RgwBucket.list(self, daemon_name=daemon['id']))
>                 users = json.loads(RgwUser.list(self, daemon_name=daemon['id']))
>                 users_count = len(users)
>                 buckets_count = len(buckets)
> ```

@Pegonzal , This code is to count the buckets/users of the daemons with no realm i.e. daemons in default zg/zone

**Comment by Pegonzal** [comment]
> > Thanks for clarifying a bit @aaSharma14, but I am still confused. Won't this code set the buckets and users to the last unique realm?
> > ```
> > else:
> >                 buckets = json.loads(RgwBucket.list(self, daemon_name=daemon['id']))
> >                 users = json.loads(RgwUser.list(self, daemon_name=daemon['id']))
> >                 users_count = len(users)
> >                 buckets_count = len(buckets)
> > ```
> 
> @Pegonzal , This code is to count the buckets/users of the daemons with no realm i.e. daemons in default zg/zone

And what will happen when:

There are 3 daemons
Realm1 has 1 bucket and 1 user.
Realm2 has 1 bucket and 1 user.

Daemon1 -> realm1 ==> adds 1 to `users_count` and `buckets_count`  since is a unique realm and adds this realm to the unique list.
Daemon2 -> realm2 ==> adds 1 to `users_count` and `buckets_count`  since is a unique realm and adds this realm to the unique list.
Daemon3 -> realm2 ==> Is not a unique realm anymore, sets to 1 `users_count` and `buckets_count`.

 **`users_count` and `buckets_count` expected value 2
  `users_count` and `buckets_count` current value 1**

**Comment by aaSharma14** [comment]
> > > Thanks for clarifying a bit @aaSharma14, but I am still confused. Won't this code set the buckets and users to the last unique realm?
> > > ```
> > > else:
> > >                 buckets = json.loads(RgwBucket.list(self, daemon_name=daemon['id']))
> > >                 users = json.loads(RgwUser.list(self, daemon_name=daemon['id']))
> > >                 users_count = len(users)
> > >                 buckets_count = len(buckets)
> > > ```
> > 
> > 
> > @Pegonzal , This code is to count the buckets/users of the daemons with no realm i.e. daemons in default zg/zone
> 
> Sorry @aaSharma14 I got confused with the code blocks, didn't read it properly, thanks for pointing it out, but anyways, what will happen when there is no `realm_name` for a daemon, the previously stored values in `users_count` and `buckets_count` will be reseted

No realm_name in a daemon will only be in the case when you don't have multi-site created and you have 1 or n number of daemons with no realm name (In this case we will count the numbers for just one daemon with no realm_name). Once we create a realm..the daemons need to be restarted and there won't be any daemon left with no realm_name..every daemon will have a realm_name in that case.

**Review by Pegonzal — APPROVED** [review]
Thanks Aashish for helping me understand the workflow.
